pkg/rebalancer/rebalancer.go

```diff
@@ -0,0 +1,242 @@
+			klog.V(4).Infof("Removing annotation %s from PersistentVolumeClaim %s/%s to avoid needlessly deleting all PVCs to rebalance a workload.", DropPVCAnnotation, pvc.Namespace, pvc.Name)
+			err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				pvc, err := r.client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(ctx, pvc.Name, metav1.GetOptions{})
+				if err != nil {
+					return err
+				}
+				delete(pvc.Annotations, DropPVCAnnotation)
+				_, err = r.client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Update(ctx, pvc, metav1.UpdateOptions{})
+				return err
+			})
+			if err != nil {
+				return false, err
+			}
+			break
+		}
+
+		klog.V(4).Infof("Rebalancing pod %s/%s.", pod.Namespace, pod.Name)
+		err := r.rebalanceWorkload(ctx, &pod, &pvc)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// If the workloads were balanced by the operator, we wait for 5 minutes
+	// before setting the status so that we don't set upgradeable=false after
+	// balancing the pods.
+	err = wait.Poll(10*time.Second, 5*time.Minute, func() (bool, error) {
+		klog.V(4).Info("Waiting until workload %s in namespace %s becomes correctly balanced.", workload.LabelSelector, workload.Namespace)
+		return r.WorkloadCorrectlyBalanced(ctx, workload)
+	})
+	return err == nil, err
+}
+
+// rebalanceWorkload rebalances the given pod on a different node. To make sure
+// that the pod will be rescheduled on a different node, this function will
+// first cordon the node in which the pod is scheduled. Then it will verify
+// whether the PVC attached to the pod needs to be deleted or not. Only pods
+// that have PVCs with a zonal topology need to be deleted since they might
+// prevent the pod to be schedule on a node that is located in a different
+// availability zone.
+func (r *Rebalancer) rebalanceWorkload(ctx context.Context, pod *v1.Pod, pvc *v1.PersistentVolumeClaim) error {
+	node, err := r.client.CoreV1().Nodes().Get(ctx, pod.Spec.NodeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	err = drain.RunCordonOrUncordon(r.drainer, node, drain.Cordon)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := drain.RunCordonOrUncordon(r.drainer, node, drain.Uncordon)
+		if err != nil {
+			klog.Errorf("Couldn't uncordon node %v: %v.", pod.Spec.NodeName, err)
+		}
+	}()
+
+	pv, err := r.client.CoreV1().PersistentVolumes().Get(ctx, pvc.Spec.VolumeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	// Do not delete the PVC if the storage provider hasn't set the topology.kubernetes.io/zone label on the PV.
+	// In most cases, when the PV isn't zonal, pods can access it from a node in a different AZ so we don't need to delete it.
+	if _, ok := pv.GetLabels()[zonalTopologyAnnotation]; ok {
+		klog.V(2).Infof("Deleting PersistentVolumeClaim %s/%s.", pvc.Namespace, pvc.Name)
+		err = r.client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(ctx, pvc.Name, metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
```suggestion
	} else {
		klog.V(2).Infof("Keeping PersistentVolumeClaim %s/%s because it doesn't have the %q label", pvc.Namespace, pvc.Name, zonalTopologyAnnotation)
	}
```